src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceCodeFixProvider.cs

```diff
@@ -0,0 +1,212 @@
+                }
+
+                break;
+            }
+        }
+    }
+
+    private static string GetTitle(ImplementInterfaceOptions options)
+    {
+        if (options.ImplementDisposePattern)
+        {
+            return options.Explicitly
+                ? FeaturesResources.Implement_interface_explicitly_with_Dispose_pattern
+                : FeaturesResources.Implement_interface_with_Dispose_pattern;
+        }
+        else if (options.Explicitly)
+        {
+            return options.OnlyRemaining
+                ? FeaturesResources.Implement_remaining_members_explicitly
+                : FeaturesResources.Implement_all_members_explicitly;
+        }
+        else if (options.Abstractly)
+        {
+            return FeaturesResources.Implement_interface_abstractly;
+        }
+        else if (options.ThroughMember != null)
+        {
+            return string.Format(FeaturesResources.Implement_interface_through_0, options.ThroughMember.Name);
+        }
+        else
+        {
+            return FeaturesResources.Implement_interface;
+        }
+    }
+
+    private static string GetEquivalenceKey(
+        IImplementInterfaceInfo state,
+        ImplementInterfaceOptions options)
+    {
+        var interfaceType = state.InterfaceTypes.First();
+        var typeName = interfaceType.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat);
+        var assemblyName = interfaceType.ContainingAssembly.Name;
+
+        // Legacy part of the equivalence key.  Kept the same to avoid test churn.
+        var codeActionTypeName = options.ImplementDisposePattern
+            ? "Microsoft.CodeAnalysis.ImplementInterface.AbstractImplementInterfaceService+ImplementInterfaceWithDisposePatternCodeAction"
+            : "Microsoft.CodeAnalysis.ImplementInterface.AbstractImplementInterfaceService+ImplementInterfaceCodeAction";
+
+        // Consider code actions equivalent if they correspond to the same interface being implemented elsewhere
+        // in the same manner.  Note: 'implement through member' means implementing the same interface through
+        // an applicable member with the same name in the destination.
+        return options.Explicitly.ToString() + ";" +
+           options.Abstractly.ToString() + ";" +
+           options.OnlyRemaining.ToString() + ":" +
+           typeName + ";" +
+           assemblyName + ";" +
+           codeActionTypeName + ";" +
+           options.ThroughMember?.Name;
+    }
+
+    private static async IAsyncEnumerable<ImplementInterfaceOptions> GetImplementOptionsAsync(
+        Document document, IImplementInterfaceInfo state, [EnumeratorCancellation] CancellationToken cancellationToken)
+    {
+        var compilation = await document.Project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+
+        if (state.MembersWithoutExplicitOrImplicitImplementationWhichCanBeImplicitlyImplemented.Length > 0)
```

Comment:
same logic that used to make all the code actions.  now it makes the 'options' at the service layer for what possible ways there are to implement. the code action will then pass this options object to the service layer to do the actual impl.